filament_scale_enhanced/__init__.py

```diff
@@ -46,19 +44,63 @@ def get_assets():
             less=["less/filament_scale.less"]
         )
 
+
     def on_startup(self, host, port):  # pylint: disable=unused-argument
-        self.hx = HX711(20, 21)
-        self.hx.set_reading_format("LSB", "MSB")
-        self.hx.reset()
-        self.hx.power_up()
-        self.t = octoprint.util.RepeatedTimer(3.0, self.check_weight)
-        self.t.start()
+        self.time = time.time()
+        
+        chip = self._settings.get(["chip"])
+        offset = float(self._settings.get(["offset"]))
+        cal_factor = float(self._settings.get(["cal_factor"]))
+
+        self._logger.info("Setting up scale, chip: [{}], offset: [{}], cal_factor: [{}]".format(chip, offset, cal_factor))
+
+        if chip == "nau7802":
+            from .scale.nau7802 import NAU7802
+
+            bus_id = int(self._settings.get(["nau_bus_id"]))
+
+            self._logger.debug("Initializing NAU7802, bus_id: [{}]".format(bus_id))
+            self.scale = NAU7802(offset, cal_factor, bus_id)
+            self._logger.debug("Initialized NAU7802, is_ready: [{}]".format(self.scale.is_ready()))
+        elif chip == "hx711":
+            from .scale.hx711 import HX711
+
+            clock_pin = self._settings.get(["hx_clockpin"])
+            data_pin = self._settings.get(["hx_datapin"])
+
+            self._logger.debug("Initializing HX711, clock_pin: [{}], data_pin: [{}]".format(clock_pin, data_pin))
+            self.scale = HX711(offset, cal_factor, clock_pin, data_pin)
+            self._logger.debug("Initialized HX711, is_ready: [{}]".format(self.scale.is_ready()))
+        else:
+            self.scale = None
+
+        if self.scale.is_ready():
+            timer_delay = float(self._settings.get(["update_delay"]))
+            self._logger.info("Setting up RepeatedTimer for check_weight(), delay: [{}]".format(timer_delay))
+            self.t = octoprint.util.RepeatedTimer(timer_delay, self.check_weight)
+            self.t.start()
+        else:
+            self.t = None
+            self._logger.warn("Skipping RepatedTimer setup for check_weight(), scale is not ready")
+
 
     def check_weight(self):
-        self.hx.power_up()
-        v = self.hx.read()
+        chip = self._settings.get(["chip"])
+        if chip == "hx711":
+            self.scale.enable()
+
+        v = self.scale.get_weight()
+        
+        self._logger.debug("weight: [{}]".format(v))
+
+        self._settings.set(["lastknownweight"], v)
+        self._settings.save()
+
         self._plugin_manager.send_plugin_message(self._identifier, v)
-        self.hx.power_down()
+
+        if chip == "hx711":
+            self.scale.disable()
```

Comment:
I can't recall why it's necessary to power up/down here for the hx11. But maybe implementing a `initial_check_weight` method, that by default calls `get_weight` and is overriden by the hx711 chip might be a touch cleaner

Something like this in the hx11 class
```python
def initial_check_weight(self) -> float:
    self.enable()
    weight = self.get_weight()
    self.disable()
    return weight
```

Reply:
This was one of the first parts of the code I dug into and this is an example of just trying to keep things as identical as possible to the original implementation before making more changes.

When I was working on the [CircuitPython HX711 GPIO bitbang driver](https://github.com/fivesixzero/CircuitPython_HX711/blob/main/hx711/hx711_gpio.py) I definitely observed some oddities around 'enable' vs 'disable' state management. In that driver's case I basically just abandoned all hope of tracking a 'powered up' vs 'powered down' state and pared the implementation down to the barest of basics, the raw ADC count read. That has its own downside - leaving the clock line low means the first read-in could provide a stale reading - but its easy enough to just read a few dozen counts, throw away the first read (or just remove outliers) before averaging.

Even with that knowledge though there's always going to be a risk of severe timing issues causing odd behavior in general, which is kinda unavoidable when bitbanging from high in the stack like this. Even with the much more lightweight CircuitPython, platforms like `nrf` and `esp32` (which do BLE/WiFi management in background threads) can see some [weird timing issues](https://github.com/fivesixzero/CircuitPython_HX711/issues/2) pop up that can totally wreck the driver's reliability. So this kind of mitigation doesn't surprise me or seem out of place, even if I'm not sure precisely why it was included.

Anyway, long story short, I think this "state toggle" should probably stick around (for now, at least). And I agree that it probably belongs in the `scale.hx711` wrapper rather than here in the main plugin code. :)